In [7]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [8]:
df=pd.read_csv("C:\\Users\\Karthik P V\\Desktop\\CP2\\labelled_data.csv")

In [9]:
df.head()

timestamp    TP2    TP3     H1  DV_pressure  Reservoirs  \
0  2020-02-01 00:00:00 -0.012  9.358  9.340       -0.024       9.358   
1  2020-02-01 00:00:10 -0.014  9.348  9.332       -0.022       9.348   
2  2020-02-01 00:00:19 -0.012  9.338  9.322       -0.022       9.338   
3  2020-02-01 00:00:29 -0.012  9.328  9.312       -0.022       9.328   
4  2020-02-01 00:00:39 -0.012  9.318  9.302       -0.022       9.318   

   Oil_temperature  Motor_current  COMP  DV_eletric  Towers  MPG  LPS  \
0           53.600         0.0400   1.0         0.0     1.0  1.0  0.0   
1           53.675         0.0400   1.0         0.0     1.0  1.0  0.0   
2           53.600         0.0425   1.0         0.0     1.0  1.0  0.0   
3           53.425         0.0400   1.0         0.0     1.0  1.0  0.0   
4           53.475         0.0400   1.0         0.0     1.0  1.0  0.0   

   Pressure_switch  Oil_level  Caudal_impulses  Airleak  
0              1.0        1.0              1.0        0  
1              1.0        1.0              1.0        0  
2              1.0        1.0              1.0        0  
3              1.0        1.0              1.0        0  
4              1.0        1.0              1.0        0

In [16]:
df.drop('timestamp' , axis=1 , inplace=True)

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1516948 entries, 0 to 1516947
Data columns (total 16 columns):
 #   Column           Non-Null Count    Dtype  
---  ------           --------------    -----  
 0   TP2              1516948 non-null  float64
 1   TP3              1516948 non-null  float64
 2   H1               1516948 non-null  float64
 3   DV_pressure      1516948 non-null  float64
 4   Reservoirs       1516948 non-null  float64
 5   Oil_temperature  1516948 non-null  float64
 6   Motor_current    1516948 non-null  float64
 7   COMP             1516948 non-null  float64
 8   DV_eletric       1516948 non-null  float64
 9   Towers           1516948 non-null  float64
 10  MPG              1516948 non-null  float64
 11  LPS              1516948 non-null  float64
 12  Pressure_switch  1516948 non-null  float64
 13  Oil_level        1516948 non-null  float64
 14  Caudal_impulses  1516948 non-null  float64
 15  Airleak          1516948 non-null  int64  
dtypes: float64(15), in

In [18]:
df[df['Airleak']==1].shape

(29954, 16)

# Failure detection

According to the company's report , we can see that there are 29,954 instances of failure.
Now lets use clustering method to predict potential failure instances

In [15]:
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler

In [19]:
scaler = StandardScaler()
df_standardized = scaler.fit_transform(df)

In [20]:
# Apply K-means clustering
kmeans = KMeans(n_clusters=2, random_state=42)
df['Cluster'] = kmeans.fit_predict(df_standardized)


df.head()

TP2    TP3     H1  DV_pressure  Reservoirs  Oil_temperature  \
0 -0.012  9.358  9.340       -0.024       9.358           53.600   
1 -0.014  9.348  9.332       -0.022       9.348           53.675   
2 -0.012  9.338  9.322       -0.022       9.338           53.600   
3 -0.012  9.328  9.312       -0.022       9.328           53.425   
4 -0.012  9.318  9.302       -0.022       9.318           53.475   

   Motor_current  COMP  DV_eletric  Towers  MPG  LPS  Pressure_switch  \
0         0.0400   1.0         0.0     1.0  1.0  0.0              1.0   
1         0.0400   1.0         0.0     1.0  1.0  0.0              1.0   
2         0.0425   1.0         0.0     1.0  1.0  0.0              1.0   
3         0.0400   1.0         0.0     1.0  1.0  0.0              1.0   
4         0.0400   1.0         0.0     1.0  1.0  0.0              1.0   

   Oil_level  Caudal_impulses  Airleak  Cluster  
0        1.0              1.0        0        1  
1        1.0              1.0        0        1  
2        1.0              1.0        0        1  
3        1.0              1.0        0        1  
4        1.0              1.0        0        1

In [22]:
df['Cluster'].value_counts()

Cluster
1    1268718
0     248230
Name: count, dtype: int64

In [24]:
df[df['Airleak']==1]['Cluster'].value_counts()

Cluster
0    29821
1      133
Name: count, dtype: int64

In [25]:
# We can observe that , for Airleak = 1 , Cluster assignment = 0

### Using clustering method , We have observed that there are 2,48,230 instances of potential Airleak Failures